### PR TITLE
Feature/3D Kalman Filter

### PR DESCRIPTION
**Pull request by LiamOS**
Have merged current main into this branch just now.

Current behaviour:
- Assumes p=0 at track end, propagates backwards only
- All reconstructed positions saved in the TMS_Track obj
- Returns single value for momentum (need to check this is reasonable)
- Can be switched on in config
- Seems to just keep running if something terrible happens (good thing?)

Please give a scan over the Reco and TreeWriter files to make sure I haven't destroyed anything unintentionally.

### CODE REVIEW

File: src/TMS_Kalman.cpp
Comment (AsaNehm) [inline src/TMS_Kalman.cpp:1]: Trusting you on this one

File: src/TMS_Kalman.h
Comment (AsaNehm) [inline src/TMS_Kalman.h:1]: Trusting you on this one

**Review by AsaNehm — APPROVED** [review]
Looks good from my side

**Comment by jdkio** [comment]
I'm testing this branch to compare with main. I ran these three tests (with a limit of 5k files):
```
# Nersc files
ConvertToTMSTree.exe /pnfs/dune/persistent/users/abooth/Production/MiniProdN1p2-v1r1/run-spill-build/output/MiniProdN1p2_NDLAr_1E19_RHC.spill/EDEPSIM_SPILLS/00000/MiniProdN1p2_NDLAr_1E19_RHC.spill.00049.EDEPSIM_SPILLS.root
# events on TMS and LAr
ConvertToTMSTree.exe /pnfs/dune/persistent/users/kleykamp/nd_production/2024-04-18_rhc_test/edep/RHC/00m/00/antineutrino.0_1713482441.edep.root
# Lar-only but not from nersc
ConvertToTMSTree.exe /pnfs/dune/persistent/users/kleykamp/nd_production/2024-05-13_lar_only_rhc/edep/RHC/00m/00/antineutrino.0_1715654453.edep.root
```
 They all process in main (tagged `kleykamp_2024-07-24_main_test`). I tried the same with this branch (tagged `kleykamp_2024-07-24_3D-Kalman-fitter_test`), and it gets stuck the second one which is TMS/LAr events: `/pnfs/dune/persistent/users/kleykamp/nd_production/2024-04-18_rhc_test/edep/RHC/00m/00/antineutrino.0_1713482441.edep.root`
The output stops at `Processing slice 1 of event number 861 / 1842`. I tried `gdb`'s `backtrace` to see where it gets stuck, and it's consistently inside various parts of `TrackMatching3D`. Eg: 
```
0x00007ffff7b1e3b4 in TMS_TrackFinder::TrackMatching3D (this=this@entry=0x16367c0 <TMS_TrackFinder::GetFinder()::Instance>)
    at TMS_Reco.cpp:1054
1054	                  || UTracks[itU].GetZ() < 11000 || UTracks[itU].GetZ() > 20000) --itU;
(gdb) bt
#0  0x00007ffff7b1e3b4 in TMS_TrackFinder::TrackMatching3D (this=this@entry=0x16367c0 <TMS_TrackFinder::GetFinder()::Instance>)
    at TMS_Reco.cpp:1054
#1  0x00007ffff7b288d6 in TMS_TrackFinder::FindTracks (this=this@entry=0x16367c0 <TMS_TrackFinder::GetFinder()::Instance>, event=...)
    at TMS_Reco.cpp:594
#2  0x0000000000408c4d in ConvertToTMSTree (filename=..., output_filename=...) at ../src/TMS_Reco.h:192
#3  0x0000000000407bd6 in main (argc=<optimized out>, argv=<optimized out>) at ConvertToTMSTree.cpp:251
```
I tried changing `Run = false` for the kalman filter and it gets stuck in the same location. I do see [some changes](https://github.com/DUNE/dune-tms/compare/kleykamp_2024-07-24_main_test...kleykamp_2024-07-24_3D-Kalman-fitter_test) in that code compared to main that are unaffected by the kalman filter, so maybe it's an issue with that logic. I'd be happy to run the test again if that helps

**Comment by LiamOS** [comment]
@AsaNehm good chance the above issue is related to the changes you made recently

**Comment by AsaNehm** [comment]
It might very well be.
@jdkio could you try whether the problem is still there if you add an if condition for all --itU and --itV for it to be above 0 (if (itV or itU > 0) --itV or --itU;) in my changed part of TMS_Reco.cpp [commits fixing matching with two hits per plane and potentially the one afterwards as well]. I had a problem with this yesterday as well and this seems to fix it for me

**Comment by jdkio** [comment]
> It might very well be. @jdkio could you try whether the problem is still there if you add an if condition for all --itU and --itV for it to be above 0 (if (itV or itU > 0) --itV or --itU;) in my changed part of TMS_Reco.cpp [commits fixing matching with two hits per plane and potentially the one afterwards as well]. I had a problem with this yesterday as well and this seems to fix it for me

I'm not sure what you mean. Would you please submit the change?

**Comment by AsaNehm** [comment]
Did that. Please try it out

**Comment by jdkio** [comment]
This works now. I was able to test all 3 files and there was no infinite loop. Test files in 
`/exp/dune/data/users/kleykamp/dune-tms/simple_processing_test/kleykamp_2024-07-31_3D-Kalman-fitter_test`

**Review by jdkio — APPROVED** [review]
Overall looks good, just some minor comments about verbosity and branch names

File: src/TMS_TreeWriter.cpp [hunk 1]
```diff
@@ -227,15 +227,20 @@ void TMS_TreeWriter::MakeBranches() {
...
-  Reco_Tree->Branch("Direction",    RecoTrackDirection,     "Direction[nTracks][3]/F");
-  Reco_Tree->Branch("EndPos",       RecoTrackEndPos,        "EndPos[nTracks][3]/F");
-  Reco_Tree->Branch("EnergyRange",  RecoTrackEnergyRange,   "EnergyRange[nTracks]/F");
-  Reco_Tree->Branch("EnergyDeposit",RecoTrackEnergyDeposit, "EnergyDeposit[nTracks]/F");
-  Reco_Tree->Branch("Length",       RecoTrackLength,        "Length[nTracks]/F");
+  Reco_Tree->Branch("nTracks",       &nTracks,               "nTracks/I");
+  Reco_Tree->Branch("nHits",         nHitsIn3DTrack,         "nHits[nTracks]/I");
+  Reco_Tree->Branch("TrackHitPos",   RecoTrackHitPos,        "TrackHitPos[nTracks][200][3]/F");
+  Reco_Tree->Branch("nKalmanNodes",  nKalmanNodes,           "nKalmanNodes[nTracks]/I");
+  Reco_Tree->Branch("KalmanPos",     RecoTrackKalmanPos,     "TrackHitPos[nTracks][200][3]/F");
```
Comment (jdkio) [inline src/TMS_TreeWriter.cpp:234]: I noticed `TrackHitPos` is repeated. Should it be
`Reco_Tree->Branch("KalmanPos",     RecoTrackKalmanPos,     "KalmanPos[nTracks][200][3]/F");`

**Comment by jdkio** [comment]
Looks like tracks start with unusual hit positions. Here are all hits on a track as an example. The first hit should be close to the second hit but the first hit has x outside the detector, and y = 0.
```
Looping over hits for track 0
(x, y, z)=(18314, 0, 15993)
(x, y, z)=(-1349.11, -1693.37, 15913)
(x, y, z)=(-1331.81, -1699.28, 15833)
(x, y, z)=(-1266.57, -1705.3, 15753)
(x, y, z)=(-1260.28, -1711.25, 15673
```
```
Looping over hits for track 0
(x, y, z)=(18314, 0, 18153)
(x, y, z)=(601.075, -1326.34, 18073)
(x, y, z)=(619.419, -1302.9, 17993)
(x, y, z)=(626.825, -1279.26, 17913)
```
The code is setting x is z_tms_end and y defaults to zero for the first hit of each track

**Comment by jdkio** [comment]
I made this comparison between PR 122 and PR 134: [comparison.pdf](https://github.com/user-attachments/files/16776472/comparison.pdf). It's looking at plots from [Tracking_Validation.cpp](https://github.com/DUNE/dune-tms/blob/main/scripts/Validation/Tracking_Validation.cpp) for this PR and what is currently main basically. Both have their pros and cons that need to be worked out. `Direction` became `StartDirection` and so the Direction and matching angle plots are bugged

I'm still concerned about hits that aren't on the track. I think that explains why we get this behavior:
image
But I do like the y comparison:
image

**Comment by AsaNehm** [comment]
Looking through the code on main there shouldn't be a problem with the shift to a PlaneByZ with coordinates instead of the plane number

**Comment by LiamOS** [comment]
Closing as this branch has become stale and isn't worth recovering. Not sure what happened and when, couldn't figure it out.

**Pull request closed without merging**